// @quintype/components/dist/es/components/social-logins/with-social-login.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import React from 'react';
import PropTypes from 'prop-types';
import { postRequest } from '../api-client';

function createSession(provider, token) {
  return postRequest("/api/login/".concat(provider), {
    token: token,
    'set-session': true
  }).json(function (r) {
    return r;
  });
}

export var WithSocialLogin =
/*#__PURE__*/
function (_React$Component) {
  _inherits(WithSocialLogin, _React$Component);

  function WithSocialLogin(props) {
    var _this;

    _classCallCheck(this, WithSocialLogin);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(WithSocialLogin).call(this, props));
    _this.serverSideLoginPath = "/login?auth-provider=".concat(_this.props.provider, "&remote-host=").concat(global.location && global.location.origin);
    return _this;
  }

  _createClass(WithSocialLogin, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.props.initialize();
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      return this.props.children({
        login: function login(props) {
          return _this2.props.socialLogin.call(_this2, props).then(function (token) {
            return createSession(_this2.props.provider, token);
          });
        },
        serverSideLoginPath: this.serverSideLoginPath
      });
    }
  }]);

  return WithSocialLogin;
}(React.Component);
WithSocialLogin.defaultProps = {
  initialize: function initialize() {},
  // function is rebound in WithSocialLogin
  socialLogin: function socialLogin() {
    window.location = this.serverSideLoginPath;
    return Promise.reject('EXPECT_REDIRECT');
  }
};
