// @quintype/components/dist/es/components/social-logins/with-linkedin-login.js
import React from 'react';
import { WithSocialLogin } from './with-social-login';

var loadLinkedInSdk = function loadLinkedInSdk(clientKey, scope) {
  if (global.IN) {
    return;
  }

  var script = document.createElement('script');
  script.src = '//platform.linkedin.com/in.js';
  script.type = 'text/javascript';
  script.async = true;
  script.defer = true;
  script.innerHTML = "\n  api_key: ".concat(clientKey, "\n");
  document.getElementsByTagName('body')[0].appendChild(script);
};

var loginWithLinkedIn = function loginWithLinkedIn(_ref) {
  var emailMandatory = _ref.emailMandatory;

  if (!global.IN || !global.IN.User) {
    return Promise.reject("NOT_LOADED");
  }

  return new Promise(function (resolve, reject) {
    IN.User.authorize(function (e) {
      if (IN.User.isAuthorized()) {
        resolve({
          "access-token": IN.ENV.auth.oauth_token,
          client: true
        });
      } else {
        reject("NOT_GRANTED");
      }
    });
  });
};

export function WithLinkedInLogin(_ref2) {
  var clientKey = _ref2.clientKey,
      children = _ref2.children,
      scope = _ref2.scope,
      emailMandatory = _ref2.emailMandatory;
  return React.createElement(WithSocialLogin, {
    provider: 'linkedin',
    initialize: function initialize() {
      return loadLinkedInSdk(clientKey, scope);
    },
    socialLogin: function socialLogin() {
      return loginWithLinkedIn({
        emailMandatory: emailMandatory
      });
    },
    children: children
  });
}
