// @quintype/components/dist/es/components/story-elements/polltype.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import React from 'react';
import { connect } from 'react-redux';
var defaultPolltypeHost = 'https://www.polltype.com';

var PolltypeBase =
/*#__PURE__*/
function (_React$Component) {
  _inherits(PolltypeBase, _React$Component);

  function PolltypeBase() {
    _classCallCheck(this, PolltypeBase);

    return _possibleConstructorReturn(this, _getPrototypeOf(PolltypeBase).apply(this, arguments));
  }

  _createClass(PolltypeBase, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.loadPolltypeJS();
    }
  }, {
    key: "loadPolltypeJS",
    value: function loadPolltypeJS() {
      var source = this.props.polltypeHost.replace(/^https:|^http:/i, '') + '/embed.js';

      if (!global._polltypeAdded) {
        global._polltypeAdded = true;
        var script = document.createElement('script');
        script.type = 'text/javascript';
        script.src = source;
        document.body.appendChild(script);
      }
    }
  }, {
    key: "render",
    value: function render() {
      return React.createElement("div", {
        "data-polltype-embed-id": this.props.id
      });
    }
  }]);

  return PolltypeBase;
}(React.Component);

function mapStateToProps(state) {
  return {
    polltypeHost: state.qt.config["polltype-host"] || defaultPolltypeHost
  };
}

var Polltype =
/*#__PURE__*/
connect(mapStateToProps, {})(PolltypeBase);
export default Polltype;
