// @quintype/components/dist/es/components/adbutler-ad.js
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import React from "react";
import PropTypes from "prop-types";
import get from "lodash/get";

function getSize(sizeMap) {
  var canUseDOM = !!(typeof window !== 'undefined' && window.document && window.document.createElement);
  var screenWidth = canUseDOM && get(window, ["screen", "width"], 440);
  if (!screenWidth) return [];
  if (screenWidth < 441) return sizeMap["mobile"];
  if (screenWidth < 992) return sizeMap["tablet"];
  return sizeMap["desktop"];
}

export function AdbutlerAd(_ref) {
  var adtype = _ref.adtype,
      adbutlerConfig = _ref.adbutlerConfig,
      sizes = _ref.sizes;
  var _adbutlerConfig$publi = adbutlerConfig.publisherId,
      publisherId = _adbutlerConfig$publi === void 0 ? "" : _adbutlerConfig$publi,
      _adbutlerConfig$adtyp = adbutlerConfig[adtype],
      zoneId = _adbutlerConfig$adtyp === void 0 ? "" : _adbutlerConfig$adtyp;
  var sizeMap = sizes[adtype];
  var size = getSize(sizeMap);

  var _size = _slicedToArray(size, 2),
      _size$ = _size[0],
      width = _size$ === void 0 ? 0 : _size$,
      _size$2 = _size[1],
      height = _size$2 === void 0 ? 0 : _size$2;

  var src = "https://servedbyadbutler.com/adserve/;ID=".concat(publisherId, ";size=").concat(width, "x").concat(height, ";setID=").concat(zoneId, ";type=iframe;click=CLICK_MACRO_PLACEHOLDER");
  var hasSize = size.length;
  return React.createElement("div", {
    className: "adbutler-wrapper"
  }, hasSize ? React.createElement("iframe", {
    src: src,
    width: width,
    height: height,
    marginWidth: "0",
    marginHeight: "0",
    hspace: "0",
    vspace: "0",
    frameBorder: "0",
    scrolling: "no"
  }) : null);
}
