// @quintype/components/dist/es/components/image-gallery.js
import React from "react";
import { ResponsiveImage } from "./responsive-image";

var ImageGalleryElement = function ImageGalleryElement(_ref) {
  var className = _ref.className,
      imageAspectRatio = _ref.imageAspectRatio,
      defaultWidth = _ref.defaultWidth,
      _ref$element = _ref.element,
      element = _ref$element === void 0 ? {} : _ref$element,
      widths = _ref.widths,
      story = _ref.story,
      onClickHandler = _ref.onClickHandler;
  var images = element['story-elements'].map(function (image, index) {
    return React.createElement("figure", {
      "data-test-id": "story-element-image-container",
      key: image.id,
      className: "story-element-image-gallery__image-container ".concat(onClickHandler ? 'custom-cursor' : ''),
      onClick: function onClick() {
        return onClickHandler && onClickHandler(index);
      }
    }, React.createElement(ResponsiveImage, {
      slug: image["image-s3-key"],
      metadata: image["metadata"],
      aspectRatio: imageAspectRatio ? imageAspectRatio : [1, 1],
      defaultWidth: defaultWidth ? defaultWidth : 480,
      widths: widths ? widths : [250, 480, 640],
      imgParams: {
        auto: ['format', 'compress']
      },
      className: className ? className : 'story-element-image-gallery__image',
      alt: image.title || story.headline
    }));
  });
  return React.createElement("div", {
    className: "story-element-image-gallery",
    "data-test-id": "story-element-image-gallery"
  }, images);
};

export { ImageGalleryElement };
