// @quintype/components/dist/es/components/link.js
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import React from "react";
import { connect } from "react-redux";

function preventDefaultImpl(e) {
  e.preventDefault();
  e.stopPropagation();
}

function LinkBase(_ref) {
  var navigateToImpl = _ref.navigateToImpl,
      externalLink = _ref.externalLink,
      callback = _ref.callback,
      href = _ref.href,
      currentHostUrl = _ref.currentHostUrl,
      _ref$navigateTo = _ref.navigateTo,
      navigateTo = _ref$navigateTo === void 0 ? navigateToImpl : _ref$navigateTo,
      _ref$preventDefault = _ref.preventDefault,
      preventDefault = _ref$preventDefault === void 0 ? preventDefaultImpl : _ref$preventDefault,
      _ref$disableAjaxLinks = _ref.disableAjaxLinks,
      disableAjaxLinks = _ref$disableAjaxLinks === void 0 ? global.disableAjaxLinks || global.disableAjaxNavigation : _ref$disableAjaxLinks,
      otherProps = _objectWithoutProperties(_ref, ["navigateToImpl", "externalLink", "callback", "href", "currentHostUrl", "navigateTo", "preventDefault", "disableAjaxLinks"]);

  return React.createElement("a", Object.assign(otherProps, {
    href: href,
    onClick: function onClick(e) {
      if (disableAjaxLinks || e.ctrlKey || e.metaKey || e.shiftKey) {
        return;
      }

      var relativeLink = href.startsWith(currentHostUrl) ? href.replace(currentHostUrl, "") : href;

      if (!relativeLink.startsWith("/")) {
        return;
      }

      preventDefault(e);

      if (externalLink) {
        global.open(externalLink, "_blank");
      } else {
        navigateTo(relativeLink);
      }

      typeof callback === 'function' && callback(e);
    }
  }));
}

function mapStateToProps(state) {
  return {
    currentHostUrl: state.qt && state.qt.currentHostUrl
  };
}

function mapDispatchToProps(dispatch) {
  return {
    navigateToImpl: function navigateToImpl(url) {
      global.app.navigateToPage(dispatch, url);
    }
  };
}

export var Link =
/*#__PURE__*/
connect(mapStateToProps, mapDispatchToProps)(LinkBase);
