// @quintype/components/dist/es/components/search-page-base.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import React from 'react';
import get from "lodash/get";
import { getRequest } from './api-client';
import { LoadMoreStoriesManager } from './load-more-stories-base';
export var SearchPageBase =
/*#__PURE__*/
function (_React$Component) {
  _inherits(SearchPageBase, _React$Component);

  function SearchPageBase() {
    _classCallCheck(this, SearchPageBase);

    return _possibleConstructorReturn(this, _getPrototypeOf(SearchPageBase).apply(this, arguments));
  }

  _createClass(SearchPageBase, [{
    key: "search",
    value: function search(pageNumber) {
      return getRequest("/api/v1/search", Object.assign(this.props.params, {
        offset: (this.props.storiesPerPage || 20) * pageNumber,
        fields: this.props.fields
      })).json(function (response) {
        return get(response, ["results", "stories"], []);
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this = this;

      return React.createElement(LoadMoreStoriesManager, Object.assign({}, this.props.data, {
        template: this.props.template,
        loadStories: function loadStories(pageNumber) {
          return _this.search(pageNumber);
        }
      }));
    }
  }]);

  return SearchPageBase;
}(React.Component);
