// @quintype/components/dist/es/components/with-preview.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import React from "react";
export function WithPreview(klazz, updateData) {
  return (
    /*#__PURE__*/
    function (_React$Component) {
      _inherits(WithPreviewWrapper, _React$Component);

      function WithPreviewWrapper(props) {
        var _this;

        _classCallCheck(this, WithPreviewWrapper);

        _this = _possibleConstructorReturn(this, _getPrototypeOf(WithPreviewWrapper).call(this, props));
        _this.state = {
          story: null
        };
        return _this;
      }

      _createClass(WithPreviewWrapper, [{
        key: "componentDidMount",
        value: function componentDidMount() {
          var _this2 = this;

          global.addEventListener("message", function (event) {
            if (event.data && event.data.action == 'reloadStory' && event.data.story) {
              _this2.setState({
                story: event.data.story
              });
            }
          });
        }
      }, {
        key: "render",
        value: function render() {
          if (!this.state.story) {
            return React.createElement("div", {
              style: {
                minHeight: 200
              }
            });
          }

          return React.createElement(klazz, Object.assign({}, this.props, {
            data: updateData(this.props.data, this.state.story)
          }));
        }
      }]);

      return WithPreviewWrapper;
    }(React.Component)
  );
}
